tests/integ/modin/frame/test_apply.py

```diff
@@ -554,33 +544,71 @@ def g(v):
         ]
 
 
-TRANSFORM_DATA_FUNC_MAP = [
-    [[[0, 1, 2], [1, 2, 3]], lambda x: x + 1],
-    [[[0, 1, 2], [1, 2, 3]], np.exp],
-    [[[0, 1, 2], [1, 2, 3]], "exp"],
-    [[["Leonhard", "Jianzhun"]], lambda x: x + " is awesome!!"],
-    [[[1.3, 2.5]], np.sqrt],
-    [[[1.3, 2.5]], "sqrt"],
-    [[[1.3, 2.5]], np.log],
-    [[[1.3, 2.5]], "log"],
-    [[[1.3, 2.5]], np.square],
-    [[[1.3, 2.5]], "square"],
+@pytest.mark.xfail(
+    strict=True,
+    raises=SnowparkSQLException,
+    reason="SNOW-1650918: Apply on dataframe data columns containing NULL fails with invalid arguments to udtf function",
+)
+@pytest.mark.parametrize(
+    "data, apply_func",
     [
-        [[None, "abcd"]],
-        lambda x: x + " are first 4 letters of alphabet" if x is not None else None,
+        [
+            [[None, "abcd"]],
+            lambda x: x + " are first 4 letters of alphabet" if x is not None else None,
+        ],
+        [
+            [[123, None]],
+            lambda x: x + 100 if x is not None else None,
+        ],
     ],
-    [[[1.5, float("nan")]], lambda x: np.sqrt(x)],
+)
+def test_apply_bug_1650918(data, apply_func):
+    native_df = native_pd.DataFrame(data)
+    snow_df = pd.DataFrame(native_df)
+    eval_snowpark_pandas_result(
+        snow_df,
+        native_df,
+        lambda x: x.apply(apply_func, axis=1),
+    )
+
+
+TRANSFORM_DATA_FUNC_MAP = [
+    [[[0, 1, 2], [1, 2, 3]], lambda x: x + 1, 16],
+    [[[0, 1, 2], [1, 2, 3]], np.exp, 16],
+    [[[0, 1, 2], [1, 2, 3]], "exp", None],
+    [[["Leonhard", "Jianzhun"]], lambda x: x + " is awesome!!", 11],
+    [[[1.3, 2.5]], np.sqrt, 11],
+    [[[1.3, 2.5]], "sqrt", None],
+    [[[1.3, 2.5]], np.log, 11],
+    [[[1.3, 2.5]], "log", None],
+    [[[1.3, 2.5]], np.square, 11],
+    [[[1.3, 2.5]], "square", None],
+    [[[1.5, float("nan")]], lambda x: np.sqrt(x), 11],
 ]
 
 
 @pytest.mark.modin_sp_precommit
-@pytest.mark.parametrize("data, apply_func", TRANSFORM_DATA_FUNC_MAP)
-@sql_count_checker(query_count=0)
-def test_basic_dataframe_transform(data, apply_func):
-    msg = "Snowpark pandas apply API doesn't yet support axis == 0"
-    with pytest.raises(NotImplementedError, match=msg):
+@pytest.mark.parametrize(
+    "data, apply_func, expected_query_count", TRANSFORM_DATA_FUNC_MAP
+)
+def test_basic_dataframe_transform(data, apply_func, expected_query_count):
+    if expected_query_count is None:
+        msg = "Snowpark pandas apply API only supports callables func"
+        with pytest.raises(NotImplementedError, match=msg):
+            snow_df = pd.DataFrame(data)
+            snow_df.transform(apply_func)
+    else:
+        msg = "SNOW-1650644 & SNOW-1345395: Avoid extra caching and repeatedly creating same temp function"
```

Comment:
I'm not that familiar with df.transform as that was not the main objective of this pr, it's just something that was enabled for free as a byproduct of supporting df.apply(axis=0).  There may be other ways to implement df.transform not sure on that, this the default implementation. 

@sfc-gh-helmeleegy Do you think we should continue failing df.transform here over performance?  The approach with df.apply(axis=0) is to reuse existing building blocks like groupby_apply so these are pre-existing perf issues that are otherwise orthogonal to this pr.  I was thinking we can make forward progress on perf issues in parallel since we also have a lot of focus there and can start getting feedback to prioritize our efforts.  Any thoughts here?

Reply:
@sfc-gh-evandenberg, would it be possible to run a small test to compare perf with and without caching for df.apply(axis=0) and df.transform()? If it becomes clear that caching is not helping in these cases (seems like it given the high number of queries), then I believe we should turn it off for those cases for now. Later, we should have a more comprehensive plan to decide on a general caching strategy for all the apply/transform use cases.

Reply:
@sfc-gh-helmeleegy I ran the related unit tests with and without caching and found they are ~ 2.5x slower if there is no caching.  Yes, let's agree to follow up here (we have tracking jiras) on more comprehensive plan for general caching.